libs/nio/src/main/java/org/elasticsearch/nio/NioSelector.java

```diff
@@ -284,24 +289,30 @@ private void handleTask(Runnable task) {
      * @param writeOperation to be queued
      */
     public void queueWrite(WriteOperation writeOperation) {
-        queuedWrites.offer(writeOperation);
-        if (isOpen() == false) {
-            boolean wasRemoved = queuedWrites.remove(writeOperation);
-            if (wasRemoved) {
-                writeOperation.getListener().accept(null, new ClosedSelectorException());
-            }
+        if (isOnCurrentThread()) {
```

Comment:
If we're going to move to this world, please remove the logic in `SocketChannelContext#sendMessage`. In the current world `queueWrite` is only called if we are not on the current thread.

Reply:
Also `writeToChannel` should not be public after this chance since there will be a single write method.

Reply:
I already did remove the logic in `SocketChannelContext#sendMessage` didn't I? (there's no more non-test users of `isOnCurrentThread` left with the current changes as far as I can see.

> Also writeToChannel should not be public after this chance since there will be a single write method.

Right, I initially thought we needed that in the tests but turns out we don't. My bad :)

Reply:
I missed the change in `SocketChannelContext#sendMessag`. Thanks for making that change.